[RFC] vim-patch:7.4.{2170,2180,2240,2241,2242}

jbradaric opened the pull request:
```
vim-patch:7.4.2170

Problem:    Cannot get information about timers.
Solution:   Add timer_info().
```
https://github.com/vim/vim/commit/8e97bd74b5377753597e3d98e7123d8985c7fffd

```
vim-patch:7.4.2180

Problem:    There is no easy way to stop all timers.  There is no way to
            temporary pause a timer.
Solution:   Add timer_stopall() and timer_pause().
```
https://github.com/vim/vim/commit/b73598e2f022a22fec512ea681c70d2775e8fd87

```
vim-patch:7.4.2240

Problem:    Tests using the sleep time can be flaky.
Solution:   Use reltime() if available. (Partly by Shane Harper)
```
https://github.com/vim/vim/commit/f267f8bdf777073e392ada5b31d837c7b6090eb4


```
vim-patch:7.4.2241

Problem:    Timer test sometimes fails.
Solution:   Increase the maximum time for repeating timer.
```
https://github.com/vim/vim/commit/973365dcc40a41e6b72ece56f15cebfee69b1329


```
vim-patch:7.4.2242

Problem:    Timer test sometimes fails.
Solution:   Increase the maximum time for callback timer test.
```
https://github.com/vim/vim/commit/17f1347b867cbcc0ce380bf9a2466b4c31896f04

Commit:
vim-patch:7.4.2170

Problem:    Cannot get information about timers.
Solution:   Add timer_info().

https://github.com/vim/vim/commit/8e97bd74b5377753597e3d98e7123d8985c7fffd

Commit:
vim-patch:7.4.2180

Problem:    There is no easy way to stop all timers.  There is no way to
            temporary pause a timer.
Solution:   Add timer_stopall() and timer_pause().

https://github.com/vim/vim/commit/b73598e2f022a22fec512ea681c70d2775e8fd87

Shougo commented:
7.4.2181 is not needed?

brcolow commented:
Not a big deal what-so-ever but the title of the PR is wrong. Should be 2170 and 2180 not 1270 and 1280. This caught my eye because I knew we had merged all patches up to a higher number 👯

jbradaric commented:
@Shougo No, we don't need 7.4.2181, `stop_all_timers()` is different in Neovim because timers are implemented in a different way. We also don't need 7.4.2171. I'll mark both patches as NA in a separate commit.

Commit:
version.c: Mark 7.4.2171 and 7.4.2181 as NA.

jbradaric commented:
7.4.2240, 7.4.2241, and 7.4.2242 also look relevant

justinmk commented:
Is this still WIP? Do you plan to add 7.4.2240, 7.4.2241, and 7.4.2242 here?

jbradaric commented:
@justinmk Yes, this is still WIP, I'll add the mentioned patches today or tomorrow.

jbradaric committed:
vim-patch:7.4.2240

Problem:    Tests using the sleep time can be flaky.
Solution:   Use reltime() if available. (Partly by Shane Harper)

https://github.com/vim/vim/commit/f267f8bdf777073e392ada5b31d837c7b6090eb4

jbradaric committed:
vim-patch:7.4.2241

Problem:    Timer test sometimes fails.
Solution:   Increase the maximum time for repeating timer.

https://github.com/vim/vim/commit/973365dcc40a41e6b72ece56f15cebfee69b1329

jbradaric committed:
vim-patch:7.4.2242

Problem:    Timer test sometimes fails.
Solution:   Increase the maximum time for callback timer test.

https://github.com/vim/vim/commit/17f1347b867cbcc0ce380bf9a2466b4c31896f04

jamessan committed:
Merge remote-tracking branch 'origin/master' into vim-7.4.2170

jamessan committed:
vim-patch:7.4.2281

Problem:    Timer test fails sometimes.
Solution:   Reduce minimum time by 1 msec.

https://github.com/vim/vim/commit/0426bae2abede764d0dd366a28663d1c6e6ab0fe

jamessan committed:
vim-patch:7.4.2304

Problem:    In a timer callback the timer itself can't be found or stopped.
            (Thinca)
Solution:   Do not remove the timer from the list, remember whether it was
            freed.

https://github.com/vim/vim/commit/417ccd7138d4d230d328de8b0d3892dd82ff1bee

jamessan committed:
vim-patch:7.4.2332

Problem:    Crash when stop_timer() is called in a callback of a callback.
            Vim hangs when the timer callback uses too much time.
Solution:   Set tr_id to -1 when a timer is to be deleted. Don't keep calling
            callbacks forever. (Ozaki Kiichi)

https://github.com/vim/vim/commit/75537a93e985ef32e6c267b06ce93629855dd983

jamessan committed:
vim-patch:7.4.2359

Problem:    Memory leak in timer_start().
Solution:   Check the right field to be NULL.

https://github.com/vim/vim/commit/26fe0d56912e42c2b16a61b2480e19ba569aee98

jamessan committed:
test_timers.vim: Adjust timing to handle difference in implementation

jamessan commented:
This has passed a couple rounds of CI now, so it should be good.  The differences seem to be due to difference choices in how the timer period and pausing are handled.  I'll open a separate issue to discuss that.

justinmk commented:
LGTM. @bfredl any comments?

bfredl commented on src/nvim/eval.c:
```diff
@@ -16648,6 +16649,77 @@ static bool set_ref_in_callback(Callback *callback, int copyID,
+    return;
+  }
+
+  if (timer->callback.type == kCallbackPartial) {
```
ok for now, but later on we should have `callback_to_typval`.

bfredl commented on src/nvim/eval.c:
```diff
@@ -16716,13 +16788,19 @@ static void f_timer_stop(typval_T *argvars, typval_T *rettv, FunPtr fptr)
 
+static void f_timer_stopall(typval_T *argvars, typval_T *unused, FunPtr fptr)
+{
+  timer_teardown();
```
nitpick: I think this function then should change name, `teardown` implies that it will only will be called at exit.

bfredl commented on src/nvim/eval.c:
```diff
@@ -16648,6 +16649,77 @@ static bool set_ref_in_callback(Callback *callback, int copyID,
+  tv_list_alloc_ret(rettv);
+  if (argvars[0].v_type != VAR_UNKNOWN) {
+    if (argvars[0].v_type != VAR_NUMBER) {
+      EMSG(_(e_number_exp));
```
I guess style issue, but I would have `return` after error and dedent the `else` block.

bfredl commented on src/nvim/eval.c:
```diff
@@ -16648,6 +16649,77 @@ static bool set_ref_in_callback(Callback *callback, int copyID,
+static void f_timer_pause(typval_T *argvars, typval_T *unused, FunPtr fptr)
+{
+  if (argvars[0].v_type != VAR_NUMBER) {
+    EMSG(_(e_number_exp));
```
ditto

bfredl commented:
Some nitpick/style issues, otherwise LGTM

jbradaric committed:
eval.c: Code style fixes
